### PR TITLE
EP to move subctl to its own repo epic

### LINKED ISSUE #89: Epic: Move subctl to its own repo
## Epic Description
subctl, soon to be a library itself, should be moved out of submariner-operator repo to its own repo. Everything specific to the subctl is moved to the subctl repo. Operator repo depends on the subctl repo. Operator repo should contain things that the operator framework generates/ needs. This is an alternative to the previous proposal of moving bundles out of the operator repo.

**Pros:**
1. both subctl and operator require different set of dependencies which might conflict will improving subctl/operator. For eg, subctl can't be `go install`-ed currently because the operator dependencies don't allow it.
2. operator framework upgrades would be simpler.
3. its logically correct to separate operator and a library as they are 2 different deliverables and provides different ways to install Submariner.
4. Testing criterias are also different. We don't need subctl related GHAs for operator and vice versa.
5. Development approach is also different. 1 is a CLI and another is a framework.
6. submariner-addon integration might be simplified.

**Cons:**
1. A new repo to maintain
2. Might add a new phase to the automated release process

## Acceptance Criteria
All of the subctl related code/GHAs are out of operator repo and in the new subctl repo.

## Definition of Done (Checklist)
* [ ] Code complete
* [ ] Relevant metrics added
* [ ] The acceptance criteria met
* [ ] Unit/e2e test added & pass
* [ ] CI jobs pass
* [ ] Deployed using cloud-prepare+subctl
* [ ] Deployed using ACM/OCM addon
* [ ] Deploy using Helm
* [ ] Deployed on supported platforms (for e.g kind, OCP on AWS, OCP on GCP)
* [ ] Run subctl verify, diagnose and gather
* [ ] Uninstall
* [ ] Troubleshooting (gather/diagnose) added
* [ ] Documentation added
* [ ] Release notes added

## Work Items
Depends on https://github.com/submariner-io/enhancements/issues/114

- [x] Create a new `subctl` repository and copy all the contents from `submariner-operator` repository.

### On `subctl` repository

- [x] https://github.com/submariner-io/subctl/issues/3

### On `submariner-operator` repository
Depends on https://github.com/submariner-io/submariner-operator/issues/2030

- [x] https://github.com/submariner-io/submariner-operator/issues/2030

### On `get.submariner.io` repository

- [x] https://github.com/submariner-io/get.submariner.io/issues/21

### On `submariner-website` repository

- [x]  https://github.com/submariner-io/submariner-website/issues/743

### Release process

- [x] https://github.com/submariner-io/releases/issues/340

### External dependencies

- [ ] Add `subctl` dependency in `submariner-addon`.

### PR DESCRIPTION
**Pull request by Jaanki**
This PR is an enhancement proposal detailing the design and
work items to move subctl to its owm repo.

Closes: #89

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

### CODE REVIEW

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Move subctl to its own repository
+
+## Summary
+
+Currently, `subctl` code lies in `submariner-operator` repository. Both subctl and operator require different set of dependencies which
+might conflict will improving subctl/operator. For eg, subctl can't be go install-ed currently because the operator dependencies don't allow
```
Comment (nyechiel) [inline submariner-operator/move-subctl-to-its-own-repo.md]: will improving?

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+it. It also makes operator upgrades difficult.
+
+## Proposal
+
+subctl, soon to be a library itself, should be moved out of submariner-operator repositoryto its own repo. Everything specific to the subctl
```
Comment (nyechiel) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
subctl, soon to be a library itself, should be moved out of submariner-operator repository to its own repo. Everything specific to the subctl
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+is moved to the new subctl repo. subctl repositorydepends on the Operator repo. Operator repositoryshould contain things that the operator
```
Comment (nyechiel) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
is moved to the new subctl repo. subctl repository depends on the Operator repo. Operator repository should contain things that the operator
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+* Operator framework upgrades would be simpler.
+* It's logically correct to separate operator and a library as they are 2 different deliverables and provides different ways to install
+Submariner.
+* Testing criterias are different. We don't need subctl related GHAs for operator and vice versa.
+* Development approach is different. One is a CLI and another is a framework.
+* submariner-addon integration might be simplified.
+
+### Cons
+
+* A new repositoryto maintain.
```
Comment (nyechiel) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* A new repository to maintain.
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+## Design details
+
+A new GitHub repository `subctl` will be created under `submariner-io` organisation. `submariner-operator` repositorycontent will be copied
```
Comment (nyechiel) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
A new GitHub repository `subctl` will be created under `submariner-io` organisation. `submariner-operator` repository content will be copied
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+to the new `subctl` repository and then unrelated content would be deleted. This will help in maintaining the history as to what content
+was moved out of `submariner-operator` repository. The `subctl` repository will preserve all the code owner information and admin rights
+from the `submariner-operator` repository. It will also follow
+[standatrd Go project layout](https://github.com/golang-standards/project-layout).
+
+## Work items
```
Comment (nyechiel) [inline submariner-operator/move-subctl-to-its-own-repo.md:39]: Can you touch on any possible user impact? If there should not be any (and this is only an internal code change really), I think it's important to explicitly state that. The goal is not to break any existing behavior.

**Review by nyechiel — COMMENTED** [review]
Thanks @Jaanki. Added a few comments.

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
+# Move subctl to its own repository
+
+## Summary
+
+Currently, `subctl` code lies in `submariner-operator` repository. Both subctl and operator require different set of dependencies which
```
Comment (skitt) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
Currently, `subctl` code lives in the `submariner-operator` repository. Both `subctl` and the Operator require different set of dependencies which
```

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 1 above)
Comment (skitt) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
might conflict with improving either or both. For example, `subctl` can't be `go install`-ed currently because the operator dependencies don't allow
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+might conflict will improving subctl/operator. For eg, subctl can't be go install-ed currently because the operator dependencies don't allow
+it. It also makes operator upgrades difficult.
+
+## Proposal
+
+subctl, soon to be a library itself, should be moved out of submariner-operator repositoryto its own repo. Everything specific to the subctl
+is moved to the new subctl repo. subctl repositorydepends on the Operator repo. Operator repositoryshould contain things that the operator
+framework generates/needs.
```
Comment (skitt) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
The proposal is to move `subctl` out of the `submariner-operator` repository to its own repository.
Everything specific to `subctl` will be moved to the new `subctl` repo.
The `subctl` project will depend on the Operator (_e.g._ for the CRDs, RBAC etc. which are defined in the Operator).
The `submariner-operator` repository will only contain artifacts which the Operator framework generates and/or needs.
```

(The library aspects are secondary and shouldn’t affect this proposal.)

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+* A new repositoryto maintain.
+
+## Design details
+
+A new GitHub repository `subctl` will be created under `submariner-io` organisation. `submariner-operator` repositorycontent will be copied
+to the new `subctl` repository and then unrelated content would be deleted. This will help in maintaining the history as to what content
+was moved out of `submariner-operator` repository. The `subctl` repository will preserve all the code owner information and admin rights
+from the `submariner-operator` repository. It will also follow
+[standatrd Go project layout](https://github.com/golang-standards/project-layout).
```
Comment (skitt) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
A new GitHub repository `subctl` will be created under the `submariner-io` organisation.
The `submariner-operator` repository content will be copied to the new `subctl` repository,
by pushing the git repository as-is, so that the full history is preserved.
Non-`subctl`-related content will then be removed from the `subctl` repository, and non-Operator-related content will be removed from the `submariner-operator` repository.
Once this is complete, there should be no duplicate content in the two repositories, other than the build framework.
The `subctl` repository will be set up with the same permissions, branch setup, and code ownership as the `submariner-operator` repository.
The `subctl` repository will also follow
[the standard Go project layout](https://github.com/golang-standards/project-layout).
```

**Review by skitt — CHANGES_REQUESTED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 2 above)
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
subctl, soon to be a library itself, should be moved out of submariner-operator repository to its own repository. Everything specific to `subctl`
```

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 3 above)
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
will be moved to the new `subctl` repository. which will depend on the Operator repository. The operator repository should contain things that the operator
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+## Proposal
+
+subctl, soon to be a library itself, should be moved out of submariner-operator repositoryto its own repo. Everything specific to the subctl
+is moved to the new subctl repo. subctl repositorydepends on the Operator repo. Operator repositoryshould contain things that the operator
+framework generates/needs.
+
+### Pros
+
+* Operator framework upgrades would be simpler.
+* It's logically correct to separate operator and a library as they are 2 different deliverables and provides different ways to install
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* It's logically correct to separate the operator and the library as they are 2 different deliverables and provides different ways to install
```
?

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+Submariner.
+* Testing criterias are different. We don't need subctl related GHAs for operator and vice versa.
+* Development approach is different. One is a CLI and another is a framework.
+* submariner-addon integration might be simplified.
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* Integration into `submariner-addon` might be simplified.
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+### Cons
+
+* A new repositoryto maintain.
+
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md:28]: Another con is more intricate connections, e.g. we'll be still realistically testing the operator with subctl for a while (if not forever), so this will pose some complications.

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+## Design details
+
+A new GitHub repository `subctl` will be created under `submariner-io` organisation. `submariner-operator` repositorycontent will be copied
+to the new `subctl` repository and then unrelated content would be deleted. This will help in maintaining the history as to what content
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]: I think dependency on submariner-operator should be listed here, and also explain why it needs to depend on operator?

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
+# Move subctl to its own repository
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]: So the new repo will be just the cmd, or just the library, or both?

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+* Create a new `subctl` repository and copy all the contents from `submariner-operator` repository.
+* Delete Operator specific code. In particular, `bundle`, `config`, `controllers`, `deploy`, `packagemanifests`, `package/*-operator*`,
+  `scripts/operator*`, `bundle.Dockerfile` directories/files.
+* Change module name in `go.mod` to `github.com/submariner-io/subctl` and add a dependency on `submariner-operator`.
+
+### On `submariner-operator` repository
+
+* Delete everything that is in `subctl` repository.
+
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md:53]: Work is also needed on `get.submariner` repository

**Review by mkolesnik — CHANGES_REQUESTED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 14 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md]: Both.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 12 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:28]: we are currently testing Operator with subctl. Now we will do the same but in new subctl repo.

Comment (skitt) [reply submariner-operator/move-subctl-to-its-own-repo.md:28]: This is a valid concern, but it’s not much different from the other repositories which also use `subctl` during their tests. The operator is perhaps more likely than the other projects to require concomitant changes to `subctl`, but we’d need to deal with that situation anyway to ensure that upgrades are always possible.

@Jaanki I think it would be useful to mention this:

```suggestion
Splitting up the operator and `subctl` will require care when implementing changes in either which affect the other.
```

**Review by skitt — CHANGES_REQUESTED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
+# Move subctl to its own repository
+
+## Summary
+
+Currently, `subctl` code lives in the `submariner-operator` repository. Both `subctl` and the Operator require different set of
+dependencies which might conflict with improving either or both. For example, `subctl` can't be `go install`-ed currently because the
+operator dependencies don't allow it. It also makes operator upgrades difficult.
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
Currently, `subctl` code lives in the `submariner-operator` repository. The Operator and `subctl` require different sets of
dependencies which might conflict with improving either or both. For example, `subctl` can't be `go install`-ed currently because the
Operator dependencies don't allow it. It also makes Operator upgrades difficult.
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+## Proposal
+
+The proposal is to move `subctl` out of the `submariner-operator` repository to its own repository. Everything specific to `subctl` will be
+moved to the new `subctl` repository. The `subctl` project will depend on the Operator (_e.g._ for the CRDs, RBAC etc. which are defined in
+the Operator). The `submariner-operator` repository will only contain artifacts which the Operator framework generates and/or needs.
+
+### Pros
+
+* Operator framework upgrades would be simpler.
+* It's logically correct to separate the operator and the library as they are 2 different deliverables and provides different ways to
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* It's logically correct to separate the Operator and the library as they are 2 different deliverables and provide different ways to
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+  install Submariner.
+* Testing criterias are different. We don't need subctl related GHAs for operator and vice versa.
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* Testing criterias are different. We don't need `subctl`-related GHAs for Operator and vice versa.
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+### On `subctl` repository
+
+* Create a new `subctl` repository and copy all the contents from `submariner-operator` repository.
+* Delete Operator specific code. In particular, `bundle`, `config`, `controllers`, `deploy`, `packagemanifests`, `package/*-operator*`,
+  `scripts/operator*`, `bundle.Dockerfile` directories/files.
+* Change module name in `go.mod` to `github.com/submariner-io/subctl` and add a dependency on `submariner-operator`.
+
+### On `submariner-operator` repository
+
+* Delete everything that is in `subctl` repository.
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* Delete the `subctl`-specific code.
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+### On `get.submariner.io` repository
+
+* Change URLs to download `subctl` binary from new `subctl` repository.
+
+### On `submariner-website` repository
+
+* Point to links to new `subctl` repository wherever applicable.
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* Point links to new `subctl` repository wherever applicable.
```

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 16 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md]: `Both a and b` is the usage usually. Its like saying `both of them require different dependencies`

Comment (dfarrell07) [reply submariner-operator/move-subctl-to-its-own-repo.md]: I don't think there's a logical difference between `both of them require different dependencies` and `they require different dependencies`, and the second one is more succinct. I'm fine with either, again IMHO they are logically equivalent.

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 21]
```diff
@@ -0,0 +1,69 @@
+# Move subctl to its own repository
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
# Move `subctl` to its own repository
```

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+The proposal is to move `subctl` out of the `submariner-operator` repository to its own repository. Everything specific to `subctl` will be
+moved to the new `subctl` repository. The `subctl` project will depend on the Operator (_e.g._ for the CRDs, RBAC etc. which are defined in
+the Operator). The `submariner-operator` repository will only contain artifacts which the Operator framework generates and/or needs.
+
+### Pros
+
+* Operator framework upgrades would be simpler.
+* It's logically correct to separate the Operator and the library as they are 2 different deliverables and provide different ways to
+  install Submariner.
+* Testing criterias are different. We don't need `subctl`-related GHAs for Operator and vice versa.
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
* Testing criteria are different. We don't need `subctl`-related GHAs for Operator and vice versa.
```
"Criteria" is already plural, "criterion" is the singular. I didn't know either, just learned this too FYI. :nerd_face: :+1:

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+
+* A new repository to maintain.
+* Splitting up the Operator and `subctl` will require care when implementing changes in either which affect the other.
+
+## Design details
+
+A new GitHub repository `subctl` will be created under the `submariner-io` organisation. The `submariner-operator` repository content will
+be copied to the new `subctl` repository, by pushing the git repository as-is, so that the full history is preserved. Non-`subctl`-related
+content will then be removed from the `subctl` repository, and non-Operator-related content will be removed from the `submariner-operator`
+repository. Once this is complete, there should be no duplicate content in the two repositories, other than the build framework. The
```
Comment (dfarrell07) [inline submariner-operator/move-subctl-to-its-own-repo.md]:
```suggestion
repository. Once this is complete, there should be no duplicate code content in the two repositories. The
```

As-is, IMHO this would imply the license, code of conduct, and everything CI (linter config, GHA...) should not be duplicated.

**Comment by dfarrell07** [comment]
There's a minor typo in the commit message FYI "its owm repo" -> "its own repo".

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 23 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md]: IMO, those thing go without saying. Each repo should have its own testing/GHAs. `other than the build framework` implies that only that part of the code could be duplicated.

**Comment by Jaanki** [comment]
> There's a minor typo in the commit message FYI "its owm repo" -> "its own repo".

I still don't see it (putting on specs :P)

**Comment by skitt** [comment]
> > There's a minor typo in the commit message FYI "its owm repo" -> "its own repo".
> 
> I still don't see it (putting on specs :P)

Screenshot from 2022-04-13 09-59-53

:smile:

**Comment by Jaanki** [comment]
> > > There's a minor typo in the commit message FYI "its owm repo" -> "its own repo".
> > 
> > 
> > I still don't see it (putting on specs :P)
> 
> Screenshot from 2022-04-13 09-59-53
> 
> smile

Oh god!!! Thanks @skitt

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+
+* Point links to new `subctl` repository wherever applicable.
+
+### Release process
+
+* Add a `subctl` phase in release automation after `submariner-operator` phase.
+
+### External dependencies
+
+* Add `subctl` dependency in `submariner-addon`.
```
Comment (vthapar) [inline submariner-operator/move-subctl-to-its-own-repo.md:68]: Why is this needed? IIRC, addon only needs operator framework and doesn't use subctl in any form.

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+
+* Create a new `subctl` repository and copy all the contents from `submariner-operator` repository.
+* Delete Operator specific code. In particular, `bundle`, `config`, `controllers`, `deploy`, `packagemanifests`, `package/*-operator*`,
+  `scripts/operator*`, `bundle.Dockerfile` directories/files.
+* Change module name in `go.mod` to `github.com/submariner-io/subctl` and add a dependency on `submariner-operator`.
+
+### On `submariner-operator` repository
+
+* Delete the `subctl`-specific code.
+
```
Comment (vthapar) [inline submariner-operator/move-subctl-to-its-own-repo.md:53]: Currently we also create a `subctl` image in operator repo which is used to run subctl from a pod within cluster. This image should also need to be created as part of the new `subctl` repo.

**Review by vthapar — DISMISSED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 24 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:68]: It uses `broker` and `globalnet` libraries which will be part of new `subctl` repo. So these will need to be changed.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 25 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:53]: Yes, everything `subctl` specific will be moved.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 24 above)
Comment (vthapar) [reply submariner-operator/move-subctl-to-its-own-repo.md:68]: Can't those libraries stay in operator? `subctl` repo should be a repo only for the CLI tool and any library it provides should be API for subctl commands e.g. `SubctlJoin` etc. Broker and globalnet libraries seem like something that subctl consumes and should be in operator. I could be wrong as haven't checked those libraries in a long time.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 12 above)
Comment (mkolesnik) [reply submariner-operator/move-subctl-to-its-own-repo.md:28]: > This is a valid concern, but it’s not much different from the other repositories which also use subctl during their tests. The operator is perhaps more likely than the other projects to require concomitant changes to subctl, but we’d need to deal with that situation anyway to ensure that upgrades are always possible.

Yes but the inter-connected-ness here is much more profound than in other repos, and the cyclical nature of things could get very confusing (esp. if a change requires a patch in both repos simultaneously)

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+The `subctl` repository will also follow [the standard Go project layout](https://github.com/golang-standards/project-layout).
+
+## Work items
+
+### On `subctl` repository
+
+* Create a new `subctl` repository and copy all the contents from `submariner-operator` repository.
+* Delete Operator specific code. In particular, `bundle`, `config`, `controllers`, `deploy`, `packagemanifests`, `package/*-operator*`,
+  `scripts/operator*`, `bundle.Dockerfile` directories/files.
+* Change module name in `go.mod` to `github.com/submariner-io/subctl` and add a dependency on `submariner-operator`.
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md:46]: Please add details about the end to end tests (CI jobs) needed in the new repo.

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+
+### On `submariner-operator` repository
+
+* Delete the `subctl`-specific code.
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md:51]: Please add details about the end to end tests (CI jobs) changes needed in the old repo.

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+### Cons
+
+* A new repository to maintain.
+* Splitting up the Operator and `subctl` will require care when implementing changes in either which affect the other.
+
+## Design details
+
+A new GitHub repository `subctl` will be created under the `submariner-io` organisation. The `submariner-operator` repository content will
+be copied to the new `subctl` repository, by pushing the git repository as-is, so that the full history is preserved. Non-`subctl`-related
+content will then be removed from the `subctl` repository, and non-Operator-related content will be removed from the `submariner-operator`
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]: This wasn't addressed

> I think dependency on submariner-operator should be listed here, and also explain why it needs to depend on operator?

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+
+* Change URLs to download `subctl` binary from new `subctl` repository.
+
+### On `submariner-website` repository
+
+* Point links to new `subctl` repository wherever applicable.
+
+### Release process
+
+* Add a `subctl` phase in release automation after `submariner-operator` phase.
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md:64]: Question is if it should go after or in parallel?

Would `subctl`'s `go.mod` depend on `operator` or vice versa?

Right now `releases` rebuilds `subctl` anyhow in order for the binaries to be available as part of the release.
So it doesn't matter in which order the releases or the different repos happen (generally).

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 12 above)
Comment (skitt) [reply submariner-operator/move-subctl-to-its-own-repo.md:28]: Which is pretty much what I said.

This makes explicit something which is already true. Currently, `subctl` and the operator are developed in the same repository, but they are consumable separately: one can (try to) use any version of `subctl` with any version of the operator. It’s simpler to develop against a single, common, repository, but splitting the repositories shouldn’t introduce any insurmountable challenges, and it will force us to actually think about changes that involve both projects.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 24 above)
Comment (skitt) [reply submariner-operator/move-subctl-to-its-own-repo.md:68]: If the operator needs a given functionality, it should be in the operator, and anything which wants to use it should get it from the operator. Anything else can be pulled up into `subctl`; I don’t think there’s much reason to keep functionality which the operator doesn’t need, in the operator — it just makes maintenance of the operator as a whole more complex.

**Review by skitt — APPROVED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+* Splitting up the Operator and `subctl` will require care when implementing changes in either which affect the other.
+
+## Design details
+
+A new GitHub repository `subctl` will be created under the `submariner-io` organisation. The `submariner-operator` repository content will
+be copied to the new `subctl` repository, by pushing the git repository as-is, so that the full history is preserved. Non-`subctl`-related
+content will then be removed from the `subctl` repository, and non-Operator-related content will be removed from the `submariner-operator`
+repository. Once this is complete, there should be no duplicate code content in the two repositories, other than the build framework. The
+`subctl` repository will be set up with the same permissions, branch setup, and code ownership as the `submariner-operator` repository.
+The `subctl` repository will also follow [the standard Go project layout](https://github.com/golang-standards/project-layout).
```
Comment (mkolesnik) [inline submariner-operator/move-subctl-to-its-own-repo.md]: I think there's enough work to be done without forcing this other refactor into this epic.

IMHO the value this will bring is questionable and can be discussed more thoroughly in a separate EP/PR.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 29 above)
Comment (skitt) [reply submariner-operator/move-subctl-to-its-own-repo.md:64]: `subctl`’s `go.mod` would depend on `submariner-operator`’s.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 30 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md]: This work is already being done as part of [Epic: Restructure/refactor submariner-operator repo](https://github.com/submariner-io/enhancements/issues/53). This epic will start after the restructure epic is finished.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 29 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:64]: Explained here https://github.com/submariner-io/enhancements/pull/99/files#diff-e948d2f8e6257d40d5f4ca0b019565c20aa0a12e731699310ab264232d4f5716R12

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 30 above)
Comment (mkolesnik) [reply submariner-operator/move-subctl-to-its-own-repo.md]: Then why list it here?

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 29 above)
Comment (mkolesnik) [reply submariner-operator/move-subctl-to-its-own-repo.md:64]: Won't that be problematic then, in general?

My main concern is if I want to add a field to a CRD, then I'd need to add it on the operator and only later to subctl, but the CRD the client in operator repo's `subctl` wont have this field (because the subctl code relies on the older version of the CRD without the field), while the server would already be expecting the field.

Right now all is happening inside one repo so it's all good, but this cylical dependency (operator -> subctl -> operator) could cause trouble.
I understand that the dependency on `subctl` from operator repo would be a weak one introduced by e2e, but it's still a sort of dependency

Comment (skitt) [reply submariner-operator/move-subctl-to-its-own-repo.md:64]: Specifically regarding CRD changes, we’d have to add the field to the operator, either with a default value or allowing empty values, and then update subctl. If we want to make the field compulsory with no default, we’d have to do it in two steps, but that would cause issues on upgrades anyway.

**Review by aswinsuryan — APPROVED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 23 above)
Comment (mkolesnik) [reply submariner-operator/move-subctl-to-its-own-repo.md]: You have not addressed CI clearly enough in this EP though

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 14 above)
Comment (mkolesnik) [reply submariner-operator/move-subctl-to-its-own-repo.md]: This should be clearly captured then

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 30 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md]: removed.

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 27 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:51]: added

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 26 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:46]: added

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 14 above)
Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md]: done

**Review by maayanf24 — APPROVED** [review]

**Review by tpantelis — APPROVED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 27 above)
Comment (mkolesnik) [reply submariner-operator/move-subctl-to-its-own-repo.md:51]: Theres more workflows like `e2e-new-subctl.yml` and `cross.yml`

Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:51]: `e2e-new-subctl.yml` is temporary to test the new subctl code in `/cmd/subctl` which is done as part of the `submariner-operator reafctor` and would be removed once the refactor is done.

Comment (Jaanki) [reply submariner-operator/move-subctl-to-its-own-repo.md:51]: Added `croos.yml`

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 29 above)
Comment (mkolesnik) [reply submariner-operator/move-subctl-to-its-own-repo.md:64]: An alternative approach is to add an "api" repo which hosts the various types (and clients) and then have both rely on that one, solving potential cyclical future nightmares.
A benefit of this is of course other projects can also import and use just the "api" repo and not be forced to import any other repo to get it.

We can leave this as an exercise for later though, as it's tangentially related to this change but would like to hear opinions about it.

**Review by mkolesnik — APPROVED** [review]

File: submariner-operator/move-subctl-to-its-own-repo.md (hunk 29 above)
Comment (skitt) [reply submariner-operator/move-subctl-to-its-own-repo.md:64]: > An alternative approach is to add an "api" repo which hosts the various types (and clients) and then have both rely on that one, solving potential cyclical future nightmares. A benefit of this is of course other projects can also import and use just the "api" repo and not be forced to import any other repo to get it.
> 
> We can leave this as an exercise for later though, as it's tangentially related to this change but would like to hear opinions about it.

We’ve done this in the past; there’s no need for a separate repo, the APIs can be published as separate Go modules. See https://github.com/submariner-io/submariner-operator/pull/1544 for example.
